File: 80. Remove Duplicates from Sorted Array II.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        '''
        original two-pointer version
        '''
        if not nums or len(nums) <= 2:
            return len(nums)
        read = write = 1
        count = 1
        while read < len(nums):
            if nums[read] == nums[read-1]:
                if count < 2:
                    count += 1
                    nums[write] = nums[read]
                    write += 1
                elif count >= 2:
                    count += 1
            else:
                count = 1
                nums[write] = nums[read]
                write += 1
            read += 1
        return write

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        '''
        a more shorter version
        '''
        i = 0
        for n in nums:
            if i < 2 or n > nums[i-2]:
                nums[i] = n
                i += 1
        return i

File: 80. Remove Duplicates from Sorted Array II.java
class Solution {
    public int removeDuplicates(int[] nums) {
        if (nums==null || nums.length<=2) {
            return nums.length;
        }
        int idx=2;
        for (int i=2; i<nums.length; i++) {
            if (nums[i] != nums[idx-2]) {
                nums[idx] = nums[i];
                idx++;
            }
        }
        return idx;
    }
}

class Solution {
    public int removeDuplicates(int[] nums) {
        if (nums==null || nums.length<=2) {
            return nums.length;
        }
        int read=1, write=0, count=1;
        while (read < nums.length) {
            if (nums[read] == nums[write] && count >= 2) {
                read++;
            } else if (nums[read] == nums[write]) {
                nums[++write] = nums[read];
                count++;
                read++;
            } else if (nums[read] != nums[write]) {
                nums[++write] = nums[read];
                count = 1;
                read++;
            }
        }
        return write+1;
    }
}